### PR TITLE
90% Code quality checks

### PR DESCRIPTION
**Pull request by kiyanwang**
This P/R introduces two tools into the project.

- PHP Codesniffer to detect code quality issues, this is configured to use a Coding Standards ruleset that I've defined in test/Standards/TalisPHP which is based on conventions we use on other projects.

- PHP MD which focuses more on code metrics and complexity/static analysis, which I've configured to use several built-in rulesets (codesize,unusedcode,naming)

Both of these tools can be tuned. 

The `ant build.xml` file has been updated to include new targets for linting (phpcs) and codecheck (phpmd) and are configured to fail the build if you have any errors ( we have many!)

This P/R also introduces two new commands:

`docker-compose run lint`
`docker-compose run codecheck`

### CODE REVIEW

**Comment by kiyanwang** [comment]
```bash
$ docker-compose run codecheck                                                                                                                                                                      

Starting redis-server: redis-server.
Unable to locate tools.jar. Expected to find it in /usr/lib/jvm/java-7-openjdk-amd64/lib/tools.jar
Buildfile: /var/talis-php/build.xml

clean:
    [mkdir] Created dir: /var/talis-php/build
    [mkdir] Created dir: /var/talis-php/build/coverage
     [exec] OK

code-check:
     [echo] Running PHPMD to detect code problems ...
     [exec] /var/talis-php/src/Talis/Babel/Client.php:16	The class Client has an overall complexity of 58 which is very high. The configured complexity threshold is 50.
     [exec] /var/talis-php/src/Talis/Babel/Client.php:215	The method createAnnotation() has a Cyclomatic Complexity of 13. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Babel/Client.php:215	The method createAnnotation() has an NPath complexity of 2048. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/Babel/Client.php:414	The method performBabelPost() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:12	The class Client has an overall complexity of 67 which is very high. The configured complexity threshold is 50.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:28	The method __construct() has an NPath complexity of 256. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:153	The method getEvents() has a Cyclomatic Complexity of 16. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:153	The method getEvents() has an NPath complexity of 6144. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:161	Avoid variables with short names like $to. Configured minimum length is 3.
     [exec] /var/talis-php/src/Talis/EchoClient/Client.php:300	The method getAnalytics() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Persona/Client/Base.php:284	The method performRequest() has a Cyclomatic Complexity of 13. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Persona/Client/Base.php:284	The method performRequest() has an NPath complexity of 1920. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/Persona/Client/Base.php:284	The method performRequest() has 110 lines of code. Current threshold is set to 100. Avoid really long methods.
     [exec] /var/talis-php/src/Talis/Persona/Client/Login.php:58	The method validateAuth() has a Cyclomatic Complexity of 15. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Persona/Client/Login.php:58	The method validateAuth() has an NPath complexity of 3074. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/Persona/Client/OAuthClients.php:40	The method updateOAuthClient() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Persona/Client/Tokens.php:15	The class Tokens has an overall complexity of 77 which is very high. The configured complexity threshold is 50.
     [exec] /var/talis-php/src/Talis/Persona/Client/Tokens.php:181	The method obtainNewToken() has a Cyclomatic Complexity of 11. The configured cyclomatic complexity threshold is 10.
     [exec] /var/talis-php/src/Talis/Persona/Client/Tokens.php:237	The method presignUrl() has an NPath complexity of 256. The configured NPath complexity threshold is 200.
     [exec] /var/talis-php/src/Talis/Persona/Client/Tokens.php:431	Avoid unused local variables such as '$exception'.
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:6	Constant Success should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:7	Constant InvalidPublicKey should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:8	Constant InvalidToken should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:9	Constant EmptyResponse should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:10	Constant Unknown should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:11	Constant Unauthorised should be defined in uppercase
     [exec] /var/talis-php/src/Talis/Persona/Client/ValidationResults.php:12	Constant InvalidSignature should be defined in uppercase

BUILD FAILED
/var/talis-php/build.xml:68: exec returned: 2

Total time: 13 seconds
```

**Comment by kiyanwang** [comment]
```bash
$ docker-compose run lint                                                                                                                                                                         

Starting redis-server: redis-server.
Unable to locate tools.jar. Expected to find it in /usr/lib/jvm/java-7-openjdk-amd64/lib/tools.jar
Buildfile: /var/talis-php/build.xml

clean:
    [mkdir] Created dir: /var/talis-php/build
    [mkdir] Created dir: /var/talis-php/build/coverage
     [exec] OK

lint:
     [echo] Running PHP Codesniffer for code quality...
     [exec] EEEEEEEEEEEEE.....EE.EEEEEEEEEEEE.EEE.EEEE 42 / 42 (100%)
     [exec]
     [exec]
     [exec]
     [exec] FILE: /var/talis-php/test/unit/Echo/EventTest.php
     [exec] ----------------------------------------------------------------------
     [exec] FOUND 13 ERRORS AFFECTING 13 LINES
     [exec] ----------------------------------------------------------------------
     [exec]    4 | ERROR | [x] Expected 1 space(s) after closing parenthesis;
     [exec]      |       |     found newline
     [exec]      |       |     (Squiz.ControlStructures.ControlSignature.SpaceAfterCloseParenthesis)
     [exec]   20 | ERROR | [x] No space found after comma in function call
     [exec]      |       |     (Generic.Functions.FunctionCallArgumentSpacing.NoSpaceAfterComma)
     [exec]   43 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   44 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   46 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   47 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   59 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   63 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays
     [exec]      |       |     (Generic.Arrays.DisallowLongArraySyntax.Found)
     [exec]   84 | ERROR | [x] Short array syntax must be used to define
     [exec]      |       |     arrays

... 
 
BUILD FAILED
/var/talis-php/build.xml:61: exec returned: 2

Total time: 13 seconds
 nadeemshabir  /vagrant/talis/talis-php   code-quality-checks                                                                                                                                                                                                  
```

**Comment by kiyanwang** [comment]
I wouldn't merge this until we address the issues its identifying, that will necessitate tuning the rules to find the right fit for us.

**Comment by markwallsgrove** [comment]
@kiyanwang nice work :) could you point me to where I can turn on the code coverage metrics?

**Comment by kiyanwang** [comment]
@markwallsgrove There is already an ant target for running the coverage report

```bash
ant coverage
```

But this produces a html file inside the container. To use this we'll need to map the host filesystem to /var/talis-php inside the container like we do in other projects. the old echo php client repo has an example of this: https://github.com/talis/echo-php-client

File: docker-compose.yml [hunk 1]
```diff
@@ -1,24 +1,38 @@
...
+    image: talis/talis-php
+    volumes:
+      - ".:/var/talis-php"
+    command: /bin/bash -c "service redis-server start && source /etc/profile.d/* && ant init"
+  lint:
+    image: talis/talis-php
+    volumes:
+      - ".:/var/talis-php"
+    command: /bin/bash -c "service redis-server start && source /etc/profile.d/* && ant lint"
+  lint-stan:
```
Comment (markwallsgrove) [inline docker-compose.yml]: Leaving this in here as it is useful & there is a follow up

File: lint [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+while true; do
```
Comment (markwallsgrove) [inline lint]: useless linting command for conintous iterations

File: src/Talis/EchoClient/Base.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Talis\EchoClient;
```
Comment (markwallsgrove) [inline src/Talis/EchoClient/Base.php]: to reduce the complexity of the classes the bunch of generic functionality was moved into this new class

File: src/Talis/Persona/Client/Base.php [hunk 4]
```diff
@@ -119,21 +112,26 @@ public function __construct(array $config)
...
-        if ($this->statsD == null) {
+        if (is_null($this->statsD)) {
             $connStr = getenv(self::STATSD_CONN);
-            if (!empty($connStr)) {
-                list($host, $port) = explode(":", $connStr);
-                $conn = new \Domnikl\Statsd\Connection\Socket($host, $port);
+
+            if (!empty($connStr) && !empty(strpos($connStr, ':'))) {
+                list($host, $port) = explode(':', $connStr);
+                $conn = new UdpSocket($host, $port);
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Base.php]: TODO: UdpSocket is a new class that belongs to the StatsD library. Which version does RL use? Is this going to cause any issues? @kiyanwang would you know?

Comment (kiyanwang) [reply src/Talis/Persona/Client/Base.php]: AFAIK RL doesn't directly use the StatsD Library

There are places in the code where RL sends stats through Echo i.e. The TalisEcho wrapper class in Tarl is used for this.

Tripod-php sends stats directly to statsd https://github.com/talis/tripod-php/blob/master/src/classes/StatsD.class.php#L89-L129  but as you can see this is handrolled and does not make use of that library

File: src/Talis/Persona/Client/Signing.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
+<?php
+namespace Talis\Persona\Client;
+
+class Signing extends Base
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Signing.php]: This class has been created to reduce the complexity in the Tokens class. For some reason the signing logic was bundled in Tokens even though they are unrelated topics.

File: src/Talis/Persona/Client/Tokens.php [hunk 6]
```diff
@@ -506,30 +381,4 @@ protected function personaObtainNewToken($url, $query)
             ]
         );
     }
-
```
Comment (markwallsgrove) [inline src/Talis/Persona/Client/Tokens.php]: moved to a trait

File: lint (hunk 2 above)
Comment (kiyanwang) [reply lint]: just to be clear this is a separate script that developer can manually call, and have a linting running whenever a file changes.  This isn't tied into ant etc.

File: src/Talis/Persona/Client/ValidationResults.php [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+class ValidationResults
 {
-    const Success = 0;
-    const InvalidPublicKey = 1;
-    const InvalidToken = 2;
-    const EmptyResponse = 3;
-    const Unknown = 4;
-    const Unauthorised = 5;
-    const InvalidSignature = 6;
+    // TODO: oops, these have changed
```
Comment (kiyanwang) [inline src/Talis/Persona/Client/ValidationResults.php]: does this need changing?

**Comment by kiyanwang** [comment]
@mark can you update the original description of this p/r to better reflect what else has changed in here.

File: src/Talis/Persona/Client/Signing.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+{
+    /**
+     * Checks the supplied config, verifies that all required parameters are present and
+     * contain a non null value;
+     *
+     * @param array $config the configuration options to validate
+     * @throws \InvalidArgumentException If the config is invalid
+     */
+    protected function checkConfig(array $config)
```
Comment (kiyanwang) [inline src/Talis/Persona/Client/Signing.php]: What is this used for? If you want anyone using Signing to be forced to provide an implementation of this then shouldn't this be an abstract class and this stub declared abstract?

I can't see that this method is ever used?  The Token's class still has an implementation of this method but doesn't refer to this Signing Class at all.

**Pull request merged**